rfcs/0109-nixpkgs-generated-code-policy.md

```diff
@@ -0,0 +1,210 @@
+
+# Motivation
+[motivation]: #motivation
+
+Nixpkgs, along with every other distro, also faces a looming crisis: new open source software is increasingly not intended to be packaged by distros at all.
+Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options is healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
+
+The lowest hanging fruit is to systematize our generated code.
+We should ensure anyone can update the generated code, which means it should be built in derivations not some ad-hoc way.
+In short, we should apply the same level of rigour that we do for packages themselves to generated code.
+
+# Detailed design
+[design]: #detailed-design
+
+## Nixpkgs
+
+1. Establish the policy that all generated code in nixpkgs must be produced by a derivation.
+   The derivation should be built by CI (so exposed as some Nixpkgs in some fashion).
+
+2. Implement script(s) for maintainers which automatically builds these derivations and vendors their results to the appropriate places.
+   Running such scripts should be sufficient to regenerated all generated code in Nixpkgs.
+
+   Greenfield tooling should not be merged unless it complies with the policy from day one.
+   Existing non-compliant tooling doesn't need to be ripped out of Nixpkgs, but the "grace period" in which is brought to compliance should be bounded.
+
+3. Ensure via CI that the vendored generated code is exactly what running the scripts produce.
+   This check should be one of the "channel blocking" CI jobs.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Impurities
+
+Many `lang2nix`-type tools have impure steps today.
+Since these tools must be only invoked inside the derivations to generate code, the impure inputs must be gotten via fixed output derivations.
+This might require changes to those tools to separate the pure work from the impuire gather steps.
+
+Additionally, as @7c6f434c point out, some upstream tooling thinks it is being pure, but the "lock files" (or similar) pinning mechanism it provides isn't up to the task for Nix's purposes.
+Quicklisp, for example, uses a "weird mix of MD5 constraints and SHA1 constraints" that isn't really up to the task.
+Another example would be using git commit hashes, which, since we don't want to download the whole history, are not good enough on their own.
```

Comment:
This isn't true. Given a commit, you can fetch the tree corresponding to that commit without fetching any of the parent commits (or their data).

Reply:
And just to make it explicit: indeed the downloaded content can be verified (the previous history is mediated in hashing by just the hash of the parent commit, which is small)

Reply:
@L-as 
Err i mean the commit hash doesn't work as a fixed output hash. Yes, it does identify what to download, but isn't sufficient to very the downloaded tree in isolation.

@7c6f434c Yes, I mentioned such Merkle inclusion proofs as future work in a comment in the IPFS RFC: https://github.com/NixOS/rfcs/pull/133/files#r957445525. But I think they are well out of scope for this one.

Any idea how I might make this clearer?

Reply:
Removing the part about downloading the whole history and saying that git commit hashes are not supported by Nix would be an improvement. If you want to give a justification why they probably won't be supported really soon, say they are more complicated to verify than normal content hashes. I just agreed that whole-history justification doesn't justify things because it's not completely true; I agree that extra complexity (and git-specificity) justifies lack of Nix support.